File: redhawk/src/base/framework/java/ossie/src/org/ossie/events/Publisher.java

package org.ossie.events;

import org.omg.CORBA.ORB;
import org.omg.CORBA.Any;
import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.CosEventChannelAdmin.*;
import org.omg.CosEventComm.*;
import org.omg.PortableServer.POA;
import org.apache.log4j.Logger;
import org.ossie.corba.utils.*;
import CF.EventChannelManagerPackage.*;
import java.lang.InterruptedException;
import java.util.concurrent.locks.*;
import java.util.concurrent.TimeUnit;

public class  Publisher  {

    public abstract class  Supplier extends PushSupplierPOA {

    };


    public class Receiver extends PushSupplierPOA {

        
        public Receiver() {};

        public boolean   get_disconnect() { return _recv_disconnect; };

        public void   reset() { _recv_disconnect = false; };
      
        public void disconnect_push_supplier () {
            _logger.debug("::disconnect_push_supplier handle disconnect_push_supplier." );
            _lock.lock();
            try{
                _recv_disconnect =  true;
                _cond.signalAll();
            }finally {
                _lock.unlock();
            }
        };


        public void wait_for_disconnect ( int wait_time, int retries )
        {
            int tries=retries;
            _lock.lock();
            try {
                while( _recv_disconnect == false ) {
                    if ( wait_time > -1 ) {
                        _logger.debug("::wait_for_disconnect.. Waiting on disconnect." );
                        boolean ret = false;
                        try {
                            ret= _cond.await( wait_time, TimeUnit.MILLISECONDS );
                        } catch( InterruptedException e ) {
                        }
                        if ( !ret && tries == -1  ) {
                            break;
                        }
                        if ( tries-- < 1 )  break;
                    }
                    else {
                        try {
                            _cond.await();

                        } catch( InterruptedException e ) {
                        }
                    }
                }
            } finally {
                _lock.unlock();
            }
        
            return;
        };

      
      protected Lock                        _lock = new ReentrantLock();
      protected Condition                   _cond = _lock.newCondition();
      protected boolean                     _recv_disconnect = true;
      protected Logger                      _logger = Logger.getLogger("Publisher.Receiver");                     

    };



  class DefaultReceiver extends Receiver {

    public DefaultReceiver ( Publisher inParent )  {
	parent=inParent;
    };

    protected Publisher parent;
  };



    //
    //  Publisher for an Event Channel
    //
    // @param channel    event channel returned from the EventChannelManager
    // @param pub   interface that is notified when a disconnect occurs
    // @param retries    number of retries to perform when trying to establish  publisher interface
    // @param retry_wait number of millisecs to wait between retries
    public Publisher( EventChannel    inChannel )
	throws OperationNotAllowed
    {
	logger = Logger.getLogger("ossie.events.Publisher");

        // if user passes a bad param then throw...
        channel = inChannel;
	if ( inChannel == null ) throw new OperationNotAllowed();

        //  local supplier object 
	disconnectReceiver = new DefaultReceiver(this);
	if ( disconnectReceiver != null ) {
            org.ossie.corba.utils.activateObject(disconnectReceiver, null);
	}

        // initialize the event channel for a publisher and the local supplier interface
        connect( );

    }


    public void  terminate() {
        logger.debug("TERMINATE - START." );
        
        // disconnect the channel
        disconnect();

        // stop our disconnectReceiver from receiving data...
        if ( disconnectReceiver != null ) {
            org.ossie.corba.utils.deactivateObject(disconnectReceiver);
        }

        // free up the resource
        proxy = null;
        disconnectReceiver=null;

        logger.debug("TERMINATE - END." );
    }


    public int disconnect() {
        return this.disconnect(10,10);
    }

    public int disconnect( int retries, int retry_wait ) {

        int retval=-1;
        int tries = retries;
        if ( proxy != null ) { 
            do {
                try {
                    proxy.disconnect_push_consumer();
                    retval=0;
                    break;
                }
                catch (COMM_FAILURE ex) {
                    logger.error( "Caught COMM_FAILURE Exception disconnecting Push Supplier! Retrying..." );
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }
                }
                tries--;
            } while(tries>0);

            if ( disconnectReceiver != null ) {
                logger.debug( "Publisher::disconnect waiting for disconnect.." );
                disconnectReceiver.wait_for_disconnect(1,3);
                logger.debug( "Publisher::disconnect received disconnect.." );
            }
                
        }

        logger.debug( "Publisher disconnected ...." );     
        return retval;
    }


    public int connect() {
        return this.connect(10,10);
    }

    public int connect( int retries, int retry_wait ) {

        int retval=-1;

        if ( channel == null ) {
            return retval;
        }

	int tries=retries;
        if ( proxy == null ) {
            SupplierAdmin admin=null;
            do {
                try {
                    admin = channel.for_suppliers ();
                    break;
                }
                catch (COMM_FAILURE ex) {
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }
                } 
                tries--;
            } while ( tries > 0);
	
            if ( admin == null ) return retval;

            tries=retries;
            do {
                try {
                    proxy = admin.obtain_push_consumer ();
                    break;
                }
                catch (COMM_FAILURE ex) {
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }
                }
                tries--;
            } while ( tries > 0 );
        }

        if ( proxy == null ) return retval;

	PushSupplier   sptr=null;
	if ( disconnectReceiver != null ) {
	    sptr = disconnectReceiver._this();
	}

        // now attach supplier to the proxy
        tries=retries;
        do {
            try {
                proxy.connect_push_supplier(sptr);
                disconnectReceiver.reset();
                retval=0;
                break;
            }
            catch(BAD_PARAM ex) {
                logger.error( "Caught BAD_PARAM exception connecting Push Supplier!");
                break;
            }
            catch(AlreadyConnected ex) {
                disconnectReceiver.reset();
                retval=0;
                logger.error("Proxy Push Consumer already connected!");
                break;
            }
            catch(COMM_FAILURE ex) {
                logger.error("Caught COMM_FAILURE exception "
                        +"connecting Push Supplier! Retrying...");
            }
            if ( retry_wait > 0 ) {
                try {
                    java.lang.Thread.sleep(retry_wait*1000);
                } catch (final InterruptedException ex) {
                }
            }
            tries--;
        } while ( tries > 0 );

        return retval;
    };


    public int    push( Any data ) {
      int retval=0;
      try {
        if (  proxy != null ) {
            proxy.push(data);
        }
        else{
          retval=-1;
        }
      }
      catch( Exception ex) {
        retval=-1;
      }

      return retval;
    }



    // handle to the Event Channel ... duplicated channel so we own this object 
    protected EventChannel              channel;

    // handle to object that publishes the event to the channel's consumers
    //protected ProxyPushConsumerOperations   proxy;
    protected ProxyPushConsumer         proxy = null;

    // handle to object that responds to disconnect messages
    protected Receiver                  disconnectReceiver = null;

    //
    // logger
    protected Logger                    logger=null;


};

File: redhawk/src/base/framework/python/ossie/events/Publisher.py
#
# This file is protected by Copyright. Please refer to the COPYRIGHT file 
# distributed with this source distribution.
# 
# This file is part of REDHAWK core.
# 
# REDHAWK core is free software: you can redistribute it and/or modify it under 
# the terms of the GNU Lesser General Public License as published by the Free 
# Software Foundation, either version 3 of the License, or (at your option) any 
# later version.
# 
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
# 
# You should have received a copy of the GNU Lesser General Public License 
# along with this program.  If not, see http://www.gnu.org/licenses/.
#

import sys
import time
import Queue
import copy
import logging
import traceback
import threading

from omniORB import any, URI, CORBA
import CosEventComm__POA
import CosEventChannelAdmin, CosEventChannelAdmin__POA


class Receiver(CosEventComm__POA.PushSupplier):
    def __init__(self):
        self._recv_disconnect = True
        self.logger = logging.getLogger("ossie.events.Publisher.Receiver")
        self._lock = threading.Lock()
        self._cond = threading.Condition(self._lock)


    def __del__(self):
        self._cond.acquire()
        self._recv_disconnect = True
        self._cond.notifyAll()
        self._cond.release()
    
    def get_disconnect(self):
        return self._recv_disconnect

    def reset(self):
        self._recv_disconnect = False

    def disconnect_push_supplier(self):
        self.logger.debug("Publisher.Reciever handle disconnect_push_supplier")
        self._cond.acquire()
        self._recv_disconnect = True
        self._cond.notifyAll()
        self._cond.release()
    
    def wait_for_disconnect( self, wait_time=-1.0, retries=-1 ):
        try:
            self._cond.acquire()
            tries = retries
            while self._recv_disconnect == False:
                self.logger.debug("Publisher.Reciever .... waiting for disconnect")
                if wait_time > -1.0:
                    self._cond.wait( wait_time/1000.0 )
                    if tries == -1:
                        break
                    tries -= 1
                    if tries < 1:
                        break;
                else:
                    self._cond.wait()
        finally:
            self._cond.release()
        

class DefaultReceiver(Receiver):
    def __init__(self,parent):
        self.parent = parent
        Receiver.__init__(self)

class Publisher:
    def __init__(self, channel ):
        self.channel = channel
        self.proxy = None
        self.logger = logging.getLogger("ossie.events.Publisher")
        self.disconnectReceiver = DefaultReceiver(self)

            
        self.connect()

    def __del__(self):
        self.logger.debug("Publisher  DTOR START")
        if self.disconnectReceiver and self.disconnectReceiver.get_disconnect() == False:
            self.logger.debug("Publisher::DTOR  DISCONNECT")
            self.disconnect()
        
        self.logger.debug("Publisher::DTOR  DEACTIVATE")
        self.disconnectReceiver=None
        self.proxy=None
        self.channel=None
        self.logger.debug("Publisher  DTOR END")


    def terminate(self):
        self.logger.debug("Publisher::terminate START")
        if self.disconnectReceiver and self.disconnectReceiver.get_disconnect() == False:
            self.logger.debug("Publisher::terminate  DISCONNECT")
            self.disconnect()
        
        self.logger.debug("Publisher::terminate  DEACTIVATE")
        self.disconnectReceiver=None
        self.proxy=None
        self.channel=None
        self.logger.debug("Publisher::terminate END")

        if self.proxy:
            for x in range(10):
                try:
                    self.proxy.disconnect_push_consumer()
                    break
                except CORBA.COMM_FAILURE:
                    pass
                time.sleep(.01)

        self.proxy = None
        self.supplier = None
        self.channel = None


    def push(self,data ):
        retval=0
        edata=data
        if not isinstance(data, CORBA.Any):
            edata = any.to_any(data)
                            
        try:
            if self.proxy != None:
                self.proxy.push(edata)
        except:
            #traceback.print_exc()
            retval=-1
        
        return retval


    def disconnect(self, retries=10, retry_wait=.01):
        retval=0
        if self.channel == None:
            return retval

        if self.proxy:
            retval=-1
            for x in range(retries):
                try:
                    self.proxy.disconnect_push_consumer()
                    retval=0
                    break
                except CORBA.COMM_FAILURE:
                    self.logger.error("Publisher ::disconnect, Caught COMM_FAILURE, Retrying.")
                    pass
                time.sleep(retry_wait)

            if self.disconnectReceiver:
                self.logger.debug("Publisher ::disconnect, Waiting for disconnect.......")
                self.disconnectReceiver.wait_for_disconnect( .01, 3)
                self.logger.debug("Publisher ::disconnect, received disconnect.......")

        return retval


    def connect(self, retries=10, retry_wait=.01):
        retval=-1
        
        if self.channel == None:
            return retval
    
        if self.proxy == None:
            self.logger.debug("Getting supplier object")
            for x in range(retries):
                try: 
                    self.logger.debug("Getting supplier object" + str(self.channel) )
                    supplier_admin = self.channel.for_suppliers()
                    break
                except CORBA.COMM_FAILURE:
                    pass
                time.sleep(retry_wait)

            for x in range(retries):
                try: 
                    self.proxy = supplier_admin.obtain_push_consumer()
                    break
                except CORBA.COMM_FAILURE:
                    pass
                time.sleep(retry_wait)

        self.logger.debug("Publisher Checking proxy...")
        if self.proxy != None:
            self.logger.debug("Publisher Connect receiver to  EventChannel ")
            for x in range(retries):
                try: 
                    self.proxy.connect_push_supplier( self.disconnectReceiver._this() )
                    self.disconnectReceiver.reset()
                    retval=0
                    break
                except CORBA.BAD_PARAM:
                    break
                except CORBA.COMM_FAILURE:
                    pass
                except CosEventChannelAdmin.AlreadyConnected:
                    self.logger.debug("Publisher Already connected to  EventChannel ")
                    self.disconnectReceiver.reset()
                    retval=0
                    break
                time.sleep(retry_wait)

        return retval
